ProcessLib/ChemistryLib/PhreeqcInterface.cpp

```diff
@@ -0,0 +1,299 @@
+            if (component != components.end())
+            {
+                switch (status)
+                {
+                    case Status::SettingSolutions:
+                        // Set component concentrations.
+                        component->setAmount(
+                            (*process_solution_iter)->get(node_id));
+                        break;
+                    case Status::UpdatingProcessSolutions:
+                        // Update solutions of component transport processes.
+                        (*process_solution_iter)
+                            ->set(node_id, component->getAmount());
+                        break;
+                }
+            }
+
+            if (pv_name == "H")
+            {
+                switch (status)
+                {
+                    case Status::SettingSolutions:
+                    {
+                        // Set pH value by hydrogen concentration.
+                        auto pH =
+                            -log10((*process_solution_iter)->get(node_id));
+                        solution.setpH(pH);
+                        break;
+                    }
+                    case Status::UpdatingProcessSolutions:
+                    {
+                        // Update hydrogen concentration by pH value.
+                        auto hydrogen_concentration =
+                            pow(10, -solution.getpH());
+                        (*process_solution_iter)
+                            ->set(node_id, hydrogen_concentration);
+                        break;
+                    }
+                }
+            }
+
+            ++process_id;
+        }
+    }
+}
+
+void phreeqcIO::writeInputsToFile()
+{
+    DBUG("Writing phreeqc inputs into file '%s'.", _input_file);
+    std::ofstream outFile(_input_file, std::ofstream::out);
+    outFile << *this;
+    outFile.close();
+}
+
+std::ofstream& operator<<(std::ofstream& outFile, phreeqcIO const& phreeqcIO)
+{
+    int const nNodes = phreeqcIO._solutions.size();
+    for (int node_id = 0; node_id < nNodes; ++node_id)
+    {
+        auto const& solution = phreeqcIO._solutions[node_id];
+        outFile << "SOLUTION " << node_id + 1 << std::endl;
+        outFile << solution << std::endl;
+
+        auto const& nodal_equilibrium_phases =
+            phreeqcIO._equilibrium_phases[node_id];
+        if (!nodal_equilibrium_phases.empty())
+        {
+            outFile << "EQUILIBRIUM_PHASES " << node_id + 1 << std::endl;
+            outFile << nodal_equilibrium_phases << std::endl;
+        }
+
+        auto const& nodal_kinetic_reactant =
+            phreeqcIO._kinetic_reactants[node_id];
+        if (!nodal_kinetic_reactant.empty())
+        {
+            outFile << "KINETICS " << node_id + 1 << std::endl;
+            outFile << nodal_kinetic_reactant;
+            outFile << "-steps " << phreeqcIO._dt << std::endl << std::endl;
+        }
+
+        if (node_id == 0)
```

Comment:
Move out of the for-loop: the if-condition body looks to be independent of the for-loop.

Reply:
Blocks of keywords "RATES" and "SELECTED_OUTPUT" have to be written out after the blocks of keywords "SOLUTION", "EQUILIBRIUM_PHASES"， and “KINETICS”. For this reason, it is put inside the loop.

Reply:
I have tested moving the block of keyword SELECTED_OUTPUT out of the for-loop. The test result shows that this move will not cause difference in result.

Reply:
Removed.